From 66188b41215496a3b891ec57f3a24acce85a3eb5 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Thu, 14 Nov 2024 11:15:27 +0530
Subject: [PATCH] UPSTREAM: CP: wifi-scripts: add multi-radio config support

Update to latest openWRT scripts upto multi-radio config support.

Link: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=04fb05914ea77db9a0be2ef82249bdfd7c80fc65

Change-Id: I9149efd1814c40f3ad49e5c7e45d64274910096b
Signed-off-by: Felix Fietkau <nbd@nbd.name>
Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 .../files/lib/netifd/wireless/mac80211.sh     | 253 +++++++++-------
 .../files/usr/share/hostap/common.uc          | 101 +++++--
 .../files/usr/share/hostap/wdev.uc            |  37 ++-
 .../network/services/hostapd/files/hostapd.uc | 285 ++++++++++++++----
 .../services/hostapd/files/wpa_supplicant.uc  |  91 ++++--
 5 files changed, 530 insertions(+), 237 deletions(-)

diff --git a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
index 11b11c411f43..9998e26ec131 100755
--- a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
+++ b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
@@ -29,7 +29,7 @@ drv_mac80211_init_device_config() {
 	config_add_string path phy 'macaddr:macaddr'
 	config_add_string tx_burst
 	config_add_string distance
-	config_add_int beacon_int chanbw frag rts
+	config_add_int radio beacon_int chanbw frag rts
 	config_add_int rxantenna txantenna txpower min_tx_power
 	config_add_int num_global_macaddr multiple_bssid
 	config_add_boolean noscan ht_coex acs_exclude_dfs background_radar
@@ -159,77 +159,79 @@ mac80211_hostapd_setup_base() {
 	chan_ofs=0
 	[ "$band" = "6g" ] && chan_ofs=1
 
-	ieee80211n=1
-	ht_capab=
-	case "$htmode" in
-		VHT20|HT20|HE20) ;;
-		HT40*|VHT40|VHT80|VHT160|HE40|HE80|HE160)
-			case "$hwmode" in
-				a)
-					case "$(( (($channel / 4) + $chan_ofs) % 2 ))" in
-						1) ht_capab="[HT40+]";;
-						0) ht_capab="[HT40-]";;
-					esac
-				;;
-				*)
-					case "$htmode" in
-						HT40+) ht_capab="[HT40+]";;
-						HT40-) ht_capab="[HT40-]";;
-						*)
-							if [ "$channel" -lt 7 ]; then
-								ht_capab="[HT40+]"
-							else
-								ht_capab="[HT40-]"
-							fi
-						;;
-					esac
-				;;
-			esac
-			[ "$auto_channel" -gt 0 ] && ht_capab="[HT40+]"
-		;;
-		*) ieee80211n= ;;
-	esac
-
-	[ -n "$ieee80211n" ] && {
-		append base_cfg "ieee80211n=1" "$N"
-
-		set_default ht_coex 0
-		append base_cfg "ht_coex=$ht_coex" "$N"
-
-		json_get_vars \
-			ldpc:1 \
-			greenfield:0 \
-			short_gi_20:1 \
-			short_gi_40:1 \
-			tx_stbc:1 \
-			rx_stbc:3 \
-			max_amsdu:1 \
-			dsss_cck_40:1
-
-		ht_cap_mask=0
-		for cap in $(iw phy "$phy" info | grep 'Capabilities:' | cut -d: -f2); do
-			ht_cap_mask="$(($ht_cap_mask | $cap))"
-		done
+	if [ "$band" != "6g" ]; then
+		ieee80211n=1
+		ht_capab=
+		case "$htmode" in
+			VHT20|HT20|HE20|EHT20) ;;
+			HT40*|VHT40|VHT80|VHT160|HE40|HE80|HE160|EHT40|EHT80|EHT160)
+				case "$hwmode" in
+					a)
+						case "$(( (($channel / 4) + $chan_ofs) % 2 ))" in
+							1) ht_capab="[HT40+]";;
+							0) ht_capab="[HT40-]";;
+						esac
+					;;
+					*)
+						case "$htmode" in
+							HT40+) ht_capab="[HT40+]";;
+							HT40-) ht_capab="[HT40-]";;
+							*)
+								if [ "$channel" -lt 7 ]; then
+									ht_capab="[HT40+]"
+								else
+									ht_capab="[HT40-]"
+								fi
+							;;
+						esac
+					;;
+				esac
+				[ "$auto_channel" -gt 0 ] && ht_capab="[HT40+]"
+			;;
+			*) ieee80211n= ;;
+		esac
 
-		cap_rx_stbc=$((($ht_cap_mask >> 8) & 3))
-		[ "$rx_stbc" -lt "$cap_rx_stbc" ] && cap_rx_stbc="$rx_stbc"
-		ht_cap_mask="$(( ($ht_cap_mask & ~(0x300)) | ($cap_rx_stbc << 8) ))"
-
-		mac80211_add_capabilities ht_capab_flags $ht_cap_mask \
-			LDPC:0x1::$ldpc \
-			GF:0x10::$greenfield \
-			SHORT-GI-20:0x20::$short_gi_20 \
-			SHORT-GI-40:0x40::$short_gi_40 \
-			TX-STBC:0x80::$tx_stbc \
-			RX-STBC1:0x300:0x100:1 \
-			RX-STBC12:0x300:0x200:1 \
-			RX-STBC123:0x300:0x300:1 \
-			MAX-AMSDU-7935:0x800::$max_amsdu \
-			DSSS_CCK-40:0x1000::$dsss_cck_40
-
-		ht_capab="$ht_capab$ht_capab_flags"
-		[ -n "$ht_capab" ] && append base_cfg "ht_capab=$ht_capab" "$N"
-	}
+		[ -n "$ieee80211n" ] && {
+			append base_cfg "ieee80211n=1" "$N"
+
+			set_default ht_coex 0
+			append base_cfg "ht_coex=$ht_coex" "$N"
+
+			json_get_vars \
+				ldpc:1 \
+				greenfield:0 \
+				short_gi_20:1 \
+				short_gi_40:1 \
+				tx_stbc:1 \
+				rx_stbc:3 \
+				max_amsdu:1 \
+				dsss_cck_40:1
+
+			ht_cap_mask=0
+			for cap in $(iw phy "$phy" info | grep 'Capabilities:' | cut -d: -f2); do
+				ht_cap_mask="$(($ht_cap_mask | $cap))"
+			done
+
+			cap_rx_stbc=$((($ht_cap_mask >> 8) & 3))
+			[ "$rx_stbc" -lt "$cap_rx_stbc" ] && cap_rx_stbc="$rx_stbc"
+			ht_cap_mask="$(( ($ht_cap_mask & ~(0x300)) | ($cap_rx_stbc << 8) ))"
+
+			mac80211_add_capabilities ht_capab_flags $ht_cap_mask \
+				LDPC:0x1::$ldpc \
+				GF:0x10::$greenfield \
+				SHORT-GI-20:0x20::$short_gi_20 \
+				SHORT-GI-40:0x40::$short_gi_40 \
+				TX-STBC:0x80::$tx_stbc \
+				RX-STBC1:0x300:0x100:1 \
+				RX-STBC12:0x300:0x200:1 \
+				RX-STBC123:0x300:0x300:1 \
+				MAX-AMSDU-7935:0x800::$max_amsdu \
+				DSSS_CCK-40:0x1000::$dsss_cck_40
+
+			ht_capab="$ht_capab$ht_capab_flags"
+			[ -n "$ht_capab" ] && append base_cfg "ht_capab=$ht_capab" "$N"
+		}
+	fi
 
 	# 802.11ac
 	enable_ac=0
@@ -238,8 +240,8 @@ mac80211_hostapd_setup_base() {
 
 	idx="$channel"
 	case "$htmode" in
-		VHT20|HE20) enable_ac=1;;
-		VHT40|HE40)
+		VHT20|HE20|EHT20) enable_ac=1;;
+		VHT40|HE40|EHT40)
 			case "$(( (($channel / 4) + $chan_ofs) % 2 ))" in
 				1) idx=$(($channel + 2));;
 				0) idx=$(($channel - 2));;
@@ -247,7 +249,7 @@ mac80211_hostapd_setup_base() {
 			enable_ac=1
 			vht_center_seg0=$idx
 		;;
-		VHT80|HE80)
+		VHT80|HE80|EHT80)
 			case "$(( (($channel / 4) + $chan_ofs) % 4 ))" in
 				1) idx=$(($channel + 6));;
 				2) idx=$(($channel + 2));;
@@ -258,7 +260,7 @@ mac80211_hostapd_setup_base() {
 			vht_oper_chwidth=1
 			vht_center_seg0=$idx
 		;;
-		VHT160|HE160)
+		VHT160|HE160|EHT160|EHT320)
 			if [ "$band" = "6g" ]; then
 				case "$channel" in
 					1|5|9|13|17|21|25|29) idx=15;;
@@ -286,15 +288,32 @@ mac80211_hostapd_setup_base() {
 
 		[ "$background_radar" -eq 1 ] && append base_cfg "enable_background_radar=1" "$N"
 	}
+
+	eht_oper_chwidth=$vht_oper_chwidth
+	eht_center_seg0=$vht_center_seg0
+
 	[ "$band" = "6g" ] && {
 		op_class=
 		case "$htmode" in
-			HE20) op_class=131;;
-			HE*) op_class=$((132 + $vht_oper_chwidth))
+			HE20|EHT20) op_class=131;;
+			EHT320)
+				case "$channel" in
+					1|5|9|13|17|21|25|29|33|37|41|45|49|53|57|61) idx=31;;
+					65|69|73|77|81|85|89|93|97|101|105|109|113|117|121|125) idx=95;;
+					129|133|137|141|145|149|153|157|161|165|169|173|177|181|185|189) idx=159;;
+					193|197|201|205|209|213|217|221) idx=191;;
+				esac
+
+				op_class=137
+				eht_center_seg0=$idx
+				eht_oper_chwidth=9
+			;;
+			HE*|EHT*) op_class=$((132 + $vht_oper_chwidth));;
 		esac
 		[ -n "$op_class" ] && append base_cfg "op_class=$op_class" "$N"
 	}
 	[ "$hwmode" = "a" ] || enable_ac=0
+	[ "$band" = "6g" ] && enable_ac=0
 
 	if [ "$enable_ac" != "0" ]; then
 		json_get_vars \
@@ -417,8 +436,10 @@ mac80211_hostapd_setup_base() {
 
 	# 802.11ax
 	enable_ax=0
+	enable_be=0
 	case "$htmode" in
 		HE*) enable_ax=1 ;;
+		EHT*) enable_ax=1; enable_be=1 ;;
 	esac
 
 	if [ "$enable_ax" != "0" ]; then
@@ -492,6 +513,14 @@ mac80211_hostapd_setup_base() {
 		append base_cfg "he_mu_edca_ac_vo_timer=255" "$N"
 	fi
 
+	if [ "$enable_be" != "0" ]; then
+		append base_cfg "ieee80211be=1" "$N"
+		[ "$hwmode" = "a" ] && {
+			append base_cfg "eht_oper_chwidth=$eht_oper_chwidth" "$N"
+			append base_cfg "eht_oper_centr_freq_seg0_idx=$eht_center_seg0" "$N"
+		}
+	fi
+
 	hostapd_prepare_device_config "$hostapd_conf_file" nl80211
 	cat >> "$hostapd_conf_file" <<EOF
 ${channel:+channel=$channel}
@@ -527,7 +556,7 @@ mac80211_hostapd_setup_bss() {
 	}
 	[ "$staidx" -gt 0 -o "$start_disabled" -eq 1 ] && append hostapd_cfg "start_disabled=1" "$N"
 
-	cat >> /var/run/hostapd-$phy.conf <<EOF
+	cat >> /var/run/hostapd-$phy$vif_phy_suffix.conf <<EOF
 $hostapd_cfg
 bssid=$macaddr
 ${default_macaddr:+#default_macaddr}
@@ -547,7 +576,7 @@ mac80211_generate_mac() {
 	local phy="$1"
 	local id="${macidx:-0}"
 
-	wdev_tool "$phy" get_macaddr id=$id num_global=$num_global_macaddr mbssid=${multiple_bssid:-0}
+	wdev_tool "$phy$phy_suffix" get_macaddr id=$id num_global=$num_global_macaddr mbssid=${multiple_bssid:-0}
 }
 
 get_board_phy_name() (
@@ -650,7 +679,7 @@ mac80211_prepare_vif() {
 		monitor) prefix=mon;;
 		esac
 
-		mac80211_set_ifname "$phy" "$prefix"
+		mac80211_set_ifname "$phy$vif_phy_suffix" "$prefix"
 	}
 
 	append active_ifnames "$ifname"
@@ -851,7 +880,12 @@ mac80211_set_vif_txpower() {
 	json_get_vars vif_txpower
 	json_select ..
 
-	[ -z "$vif_txpower" ] || iw dev "$ifname" set txpower fixed "${vif_txpower%%.*}00"
+	set_default vif_txpower "$txpower"
+	if [ -n "$vif_txpower" ]; then
+		iw dev "$ifname" set txpower fixed "${vif_txpower%%.*}00"
+	else
+		iw dev "$ifname" set txpower auto
+	fi
 }
 
 wpa_supplicant_init_config() {
@@ -891,11 +925,13 @@ wpa_supplicant_add_interface() {
 
 wpa_supplicant_set_config() {
 	local phy="$1"
+	local radio="$2"
 	local prev
 
 	json_set_namespace wpa_supp prev
 	json_close_array
 	json_add_string phy "$phy"
+	json_add_int radio "$radio"
 	json_add_int num_global_macaddr "$num_global_macaddr"
 	json_add_boolean defer 1
 	local data="$(json_dump)"
@@ -918,13 +954,16 @@ wpa_supplicant_set_config() {
 }
 
 hostapd_set_config() {
+	local phy="$1"
+	local radio="$2"
+
 	[ -n "$hostapd_ctrl" ] || {
-		ubus_call hostapd config_set '{ "phy": "'"$phy"'", "config": "", "prev_config": "'"${hostapd_conf_file}.prev"'" }' > /dev/null
+		ubus_call hostapd config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "config": "", "prev_config": "'"${hostapd_conf_file}.prev"'" }' > /dev/null
 		return 0;
 	}
 
 	ubus wait_for hostapd
-	local hostapd_res="$(ubus_call hostapd config_set "{ \"phy\": \"$phy\", \"config\":\"${hostapd_conf_file}\", \"prev_config\": \"${hostapd_conf_file}.prev\"}")"
+	local hostapd_res="$(ubus_call hostapd config_set "{ \"phy\": \"$phy\", \"radio\": $radio, \"config\":\"${hostapd_conf_file}\", \"prev_config\": \"${hostapd_conf_file}.prev\"}")"
 	ret="$?"
 	[ "$ret" != 0 -o -z "$hostapd_res" ] && {
 		wireless_setup_failed HOSTAPD_START_FAILED
@@ -936,10 +975,11 @@ hostapd_set_config() {
 
 wpa_supplicant_start() {
 	local phy="$1"
+	local radio="$2"
 
 	[ -n "$wpa_supp_init" ] || return 0
 
-	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "num_global_macaddr": '"$num_global_macaddr"' }' > /dev/null
+	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "num_global_macaddr": '"$num_global_macaddr"' }' > /dev/null
 }
 
 mac80211_setup_supplicant() {
@@ -1044,18 +1084,23 @@ drv_mac80211_cleanup() {
 }
 
 mac80211_reset_config() {
-	local phy="$1"
+	hostapd_conf_file="/var/run/hostapd-$phy$vif_phy_suffix.conf"
+	ubus_call hostapd config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "config": "", "prev_config": "'"$hostapd_conf_file"'" }' > /dev/null
+	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "config": [] }' > /dev/null
+	wdev_tool "$phy$phy_suffix" set_config '{}'
+}
 
-	hostapd_conf_file="/var/run/hostapd-$phy.conf"
-	ubus_call hostapd config_set '{ "phy": "'"$phy"'", "config": "", "prev_config": "'"$hostapd_conf_file"'" }' > /dev/null
-	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "config": [] }' > /dev/null
-	wdev_tool "$phy" set_config '{}'
+mac80211_set_suffix() {
+	[ "$radio" = "-1" ] && radio=
+	phy_suffix="${radio:+:$radio}"
+	vif_phy_suffix="${radio:+.$radio}"
+	set_default radio -1
 }
 
 drv_mac80211_setup() {
 	json_select config
 	json_get_vars \
-		phy macaddr path \
+		radio phy macaddr path \
 		country chanbw distance \
 		txpower \
 		rxantenna txantenna \
@@ -1065,6 +1110,8 @@ drv_mac80211_setup() {
 	json_get_values scan_list scan_list
 	json_select ..
 
+	mac80211_set_suffix
+
 	json_select data && {
 		json_get_var prev_rxantenna rxantenna
 		json_get_var prev_txantenna txantenna
@@ -1091,7 +1138,7 @@ drv_mac80211_setup() {
 		}
 	}
 
-	hostapd_conf_file="/var/run/hostapd-$phy.conf"
+	hostapd_conf_file="/var/run/hostapd-$phy$vif_phy_suffix.conf"
 
 	macidx=0
 	staidx=0
@@ -1110,17 +1157,11 @@ drv_mac80211_setup() {
 	[ "$rxantenna" = "all" ] && rxantenna=0xffffffff
 
 	[ "$rxantenna" = "$prev_rxantenna" -a "$txantenna" = "$prev_txantenna" ] || mac80211_reset_config "$phy"
-	wireless_set_data phy="$phy" txantenna="$txantenna" rxantenna="$rxantenna"
+	wireless_set_data phy="$phy" radio="$radio" txantenna="$txantenna" rxantenna="$rxantenna"
 
 	iw phy "$phy" set antenna $txantenna $rxantenna >/dev/null 2>&1
 	iw phy "$phy" set distance "$distance" >/dev/null 2>&1
 
-	if [ -n "$txpower" ]; then
-		iw phy "$phy" set txpower fixed "${txpower%%.*}00"
-	else
-		iw phy "$phy" set txpower auto
-	fi
-
 	[ -n "$frag" ] && iw phy "$phy" set frag "${frag%%.*}"
 	[ -n "$rts" ] && iw phy "$phy" set rts "${rts%%.*}"
 
@@ -1148,13 +1189,13 @@ drv_mac80211_setup() {
 	for_each_interface "ap sta adhoc mesh monitor" mac80211_prepare_vif
 	for_each_interface "ap sta adhoc mesh monitor" mac80211_setup_vif
 
-	[ -x /usr/sbin/wpa_supplicant ] && wpa_supplicant_set_config "$phy"
-	[ -x /usr/sbin/hostapd ] && hostapd_set_config "$phy"
+	[ -x /usr/sbin/wpa_supplicant ] && wpa_supplicant_set_config "$phy" "$radio"
+	[ -x /usr/sbin/hostapd ] && hostapd_set_config "$phy" "$radio"
 
-	[ -x /usr/sbin/wpa_supplicant ] && wpa_supplicant_start "$phy"
+	[ -x /usr/sbin/wpa_supplicant ] && wpa_supplicant_start "$phy" "$radio"
 
 	json_set_namespace wdev_uc prev
-	wdev_tool "$phy" set_config "$(json_dump)" $active_ifnames
+	wdev_tool "$phy$phy_suffix" set_config "$(json_dump)" $active_ifnames
 	json_set_namespace "$prev"
 
 	for_each_interface "ap sta adhoc mesh monitor" mac80211_set_vif_txpower
@@ -1181,19 +1222,15 @@ list_phy_interfaces() {
 
 drv_mac80211_teardown() {
 	json_select data
-	json_get_vars phy
+	json_get_vars phy radio
 	json_select ..
 	[ -n "$phy" ] || {
 		echo "Bug: PHY is undefined for device '$1'"
 		return 1
 	}
 
+	mac80211_set_suffix
 	mac80211_reset_config "$phy"
-
-	for wdev in $(list_phy_interfaces "$phy"); do
-		ip link set dev "$wdev" down
-		iw dev "$wdev" del
-	done
 }
 
 add_driver mac80211
diff --git a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
index 750e3ae71c27..c6bfb62ef3c5 100644
--- a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
+++ b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
@@ -1,6 +1,6 @@
 import * as nl80211 from "nl80211";
 import * as rtnl from "rtnl";
-import { readfile, glob, basename, readlink } from "fs";
+import { readfile, glob, basename, readlink, open } from "fs";
 
 const iftypes = {
 	ap: nl80211.const.NL80211_IFTYPE_AP,
@@ -74,6 +74,14 @@ function find_reusable_wdev(phyidx)
 	return null;
 }
 
+function wdev_set_radio_mask(name, mask)
+{
+	nl80211.request(nl80211.const.NL80211_CMD_SET_INTERFACE, 0, {
+		dev: name,
+		vif_radio_mask: mask
+	});
+}
+
 function wdev_create(phy, name, data)
 {
 	let phyidx = int(readfile(`/sys/class/ieee80211/${phy}/index`));
@@ -93,24 +101,24 @@ function wdev_create(phy, name, data)
 		req["4addr"] = data["4addr"];
 	if (data.macaddr)
 		req.mac = data.macaddr;
+	if (data.radio != null && data.radio >= 0)
+		req.vif_radio_mask = 1 << data.radio;
 
 	nl80211.error();
 
 	let reuse_ifname = find_reusable_wdev(phyidx);
 	if (reuse_ifname &&
 	    (reuse_ifname == name ||
-	     rtnl.request(rtnl.const.RTM_SETLINK, 0, { dev: reuse_ifname, ifname: name}) != false))
-		nl80211.request(
-			nl80211.const.NL80211_CMD_SET_INTERFACE, 0, {
-				wiphy: phyidx,
-				dev: name,
-				iftype: iftypes[data.mode],
-			});
-	else
+	     rtnl.request(rtnl.const.RTM_SETLINK, 0, { dev: reuse_ifname, ifname: name}) != false)) {
+		req.dev = req.ifname;
+		delete req.ifname;
+		nl80211.request(nl80211.const.NL80211_CMD_SET_INTERFACE, 0, req);
+	} else {
 		nl80211.request(
 			nl80211.const.NL80211_CMD_NEW_INTERFACE,
 			nl80211.const.NLM_F_CREATE,
 			req);
+	}
 
 	let error = nl80211.error();
 	if (error)
@@ -190,7 +198,8 @@ const phy_proto = {
 	},
 
 	macaddr_generate: function(data) {
-		let phy = this.name;
+		let phy = this.phy;
+		let radio_idx = this.radio;
 		let idx = int(data.id ?? 0);
 		let mbssid = int(data.mbssid ?? 0) > 0;
 		let num_global = int(data.num_global ?? 1);
@@ -200,22 +209,30 @@ const phy_proto = {
 		if (!base_addr)
 			return null;
 
-		if (!idx && !mbssid)
-			return base_addr;
-
 		let base_mask = phy_sysfs_file(phy, "address_mask");
 		if (!base_mask)
 			return null;
 
-		if (base_mask == "00:00:00:00:00:00" && idx >= num_global) {
+		if (base_mask == "00:00:00:00:00:00" &&
+		    (radio_idx > 0 || idx >= num_global)) {
 			let addrs = split(phy_sysfs_file(phy, "addresses"), "\n");
 
-			if (idx < length(addrs))
-				return addrs[idx];
+			if (radio_idx != null) {
+				if (radio_idx && radio_idx < length(addrs))
+					base_addr = addrs[radio_idx];
+				else
+					idx += radio_idx * 16;
+			} else {
+				if (idx < length(addrs))
+					return addrs[idx];
 
-			base_mask = "ff:ff:ff:ff:ff:ff";
+				base_mask = "ff:ff:ff:ff:ff:ff";
+			}
 		}
 
+		if (!idx && !mbssid)
+			return base_addr;
+
 		let addr = macaddr_split(base_addr);
 		let mask = macaddr_split(base_mask);
 		let type;
@@ -275,27 +292,55 @@ const phy_proto = {
 		}
 	},
 
+	wdev_add: function(name, data) {
+		let phydev = this;
+		wdev_create(this.phy, name, {
+			...data,
+			radio: this.radio,
+		});
+	},
+
 	for_each_wdev: function(cb) {
-		let wdevs = glob(`/sys/class/ieee80211/${this.name}/device/net/*`);
-		wdevs = map(wdevs, (arg) => basename(arg));
+		let wdevs = nl80211.request(
+			nl80211.const.NL80211_CMD_GET_INTERFACE,
+			nl80211.const.NLM_F_DUMP,
+			{ wiphy: this.idx }
+		);
+
+		let mac_wdev = {};
 		for (let wdev in wdevs) {
-			if (basename(readlink(`/sys/class/net/${wdev}/phy80211`)) != this.name)
+			if (wdev.iftype == nl80211.const.NL80211_IFTYPE_AP_VLAN)
+				continue;
+			if (this.radio != null && wdev.vif_radio_mask != null &&
+			    !(wdev.vif_radio_mask & (1 << this.radio)))
 				continue;
+			mac_wdev[wdev.mac] = wdev;
+		}
 
-			cb(wdev);
+		for (let wdev in wdevs) {
+			if (!mac_wdev[wdev.mac])
+				continue;
+
+			cb(wdev.ifname);
 		}
 	}
 };
 
-function phy_open(phy)
+function phy_open(phy, radio)
 {
 	let phyidx = readfile(`/sys/class/ieee80211/${phy}/index`);
 	if (!phyidx)
 		return null;
 
+	let name = phy;
+	if (radio === "" || radio < 0)
+		radio = null;
+	if (radio != null)
+		name += "." + radio;
+
 	return proto({
-		name: phy,
-		idx: int(phyidx)
+		phy, name, radio,
+		idx: int(phyidx),
 	}, phy_proto);
 }
 
@@ -365,9 +410,9 @@ function is_equal(val1, val2) {
 
 function vlist_new(cb) {
 	return proto({
-			cb: cb,
-			data: {}
-		}, vlist_proto);
+		cb: cb,
+		data: {}
+	}, vlist_proto);
 }
 
-export { wdev_remove, wdev_create, wdev_set_mesh_params, wdev_set_up, is_equal, vlist_new, phy_is_fullmac, phy_open };
+export { wdev_remove, wdev_create, wdev_set_mesh_params, wdev_set_radio_mask, wdev_set_up, is_equal, vlist_new, phy_is_fullmac, phy_open };
diff --git a/package/network/config/wifi-scripts/files/usr/share/hostap/wdev.uc b/package/network/config/wifi-scripts/files/usr/share/hostap/wdev.uc
index d505e3035ffb..d836a930fc4f 100644
--- a/package/network/config/wifi-scripts/files/usr/share/hostap/wdev.uc
+++ b/package/network/config/wifi-scripts/files/usr/share/hostap/wdev.uc
@@ -1,13 +1,13 @@
 #!/usr/bin/env ucode
 'use strict';
-import { vlist_new, is_equal, wdev_create, wdev_set_mesh_params, wdev_remove, wdev_set_up, phy_open } from "/usr/share/hostap/common.uc";
+import { vlist_new, is_equal, wdev_set_mesh_params, wdev_remove, wdev_set_up, phy_open } from "/usr/share/hostap/common.uc";
 import { readfile, writefile, basename, readlink, glob } from "fs";
 let libubus = require("ubus");
 
 let keep_devices = {};
-let phy = shift(ARGV);
+let phy_name = shift(ARGV);
 let command = shift(ARGV);
-let phydev;
+let phy, phydev;
 
 function iface_stop(wdev)
 {
@@ -30,7 +30,7 @@ function iface_start(wdev)
 		wdev_config[key] = wdev[key];
 	if (!wdev_config.macaddr && wdev.mode != "monitor")
 		wdev_config.macaddr = phydev.macaddr_next();
-	wdev_create(phy, ifname, wdev_config);
+	phydev.wdev_add(ifname, wdev_config);
 	wdev_set_up(ifname, true);
 	let htmode = wdev.htmode || "NOHT";
 	if (wdev.freq)
@@ -85,20 +85,15 @@ function delete_ifname(config)
 		delete config[key].ifname;
 }
 
-function add_existing(phy, config)
+function add_existing(phydev, config)
 {
-	let wdevs = glob(`/sys/class/ieee80211/${phy}/device/net/*`);
-	wdevs = map(wdevs, (arg) => basename(arg));
-	for (let wdev in wdevs) {
+	phydev.for_each_wdev((wdev) => {
 		if (config[wdev])
-			continue;
-
-		if (basename(readlink(`/sys/class/net/${wdev}/phy80211`)) != phy)
-			continue;
+			return;
 
 		if (trim(readfile(`/sys/class/net/${wdev}/operstate`)) == "down")
 			config[wdev] = {};
-	}
+	});
 }
 
 function usage()
@@ -114,7 +109,7 @@ Commands:
 
 const commands = {
 	set_config: function(args) {
-		let statefile = `/var/run/wdev-${phy}.json`;
+		let statefile = `/var/run/wdev-${phy_name}.json`;
 
 		let new_config = shift(args);
 		for (let dev in ARGV)
@@ -137,12 +132,12 @@ const commands = {
 		if (type(old_config) == "object")
 			config.data = old_config;
 
-		add_existing(phy, config.data);
+		add_existing(phydev, config.data);
 		add_ifname(config.data);
 		drop_inactive(config.data);
 
 		let ubus = libubus.connect();
-		let data = ubus.call("hostapd", "config_get_macaddr_list", { phy: phy });
+		let data = ubus.call("hostapd", "config_get_macaddr_list", { phy: phydev.name, radio: phydev.radio ?? -1 });
 		let macaddr_list = [];
 		if (type(data) == "object" && data.macaddr)
 			macaddr_list = data.macaddr;
@@ -166,7 +161,7 @@ const commands = {
 
 		let macaddr = phydev.macaddr_generate(data);
 		if (!macaddr) {
-			warn(`Could not get MAC address for phy ${phy}\n`);
+			warn(`Could not get MAC address for phy ${phy_name}\n`);
 			exit(1);
 		}
 
@@ -174,12 +169,14 @@ const commands = {
 	},
 };
 
-if (!phy || !command | !commands[command])
+if (!phy_name || !command | !commands[command])
 	usage();
 
-phydev = phy_open(phy);
+let phy_split = split(phy_name, ":");
+phydev = phy_open(phy_split[0], phy_split[1]);
+phy = phydev.phy;
 if (!phydev) {
-	warn(`PHY ${phy} does not exist\n`);
+	warn(`PHY ${phy_name} does not exist\n`);
 	exit(1);
 }
 
diff --git a/package/network/services/hostapd/files/hostapd.uc b/package/network/services/hostapd/files/hostapd.uc
index dfddf8185ba8..2d9ce2874901 100644
--- a/package/network/services/hostapd/files/hostapd.uc
+++ b/package/network/services/hostapd/files/hostapd.uc
@@ -1,6 +1,6 @@
 let libubus = require("ubus");
 import { open, readfile } from "fs";
-import { wdev_create, wdev_remove, is_equal, vlist_new, phy_is_fullmac, phy_open } from "common";
+import { wdev_remove, is_equal, vlist_new, phy_is_fullmac, phy_open, wdev_set_radio_mask } from "common";
 
 let ubus = libubus.connect(null, 60);
 
@@ -22,6 +22,31 @@ hostapd.data.file_fields = {
 	eap_sim_db: true,
 };
 
+hostapd.data.iface_fields = {
+	ft_iface: true,
+	upnp_iface: true,
+	snoop_iface: true,
+	bridge: true,
+	iapp_interface: true,
+};
+
+hostapd.data.bss_info_fields = {
+	// radio
+	hw_mode: true,
+	channel: true,
+	ieee80211ac: true,
+	ieee80211ax: true,
+
+	// bss
+	bssid: true,
+	ssid: true,
+	wpa: true,
+	wpa_key_mgmt: true,
+	wpa_pairwise: true,
+	auth_algs: true,
+	ieee80211w: true,
+};
+
 function iface_remove(cfg)
 {
 	if (!cfg || !cfg.bss || !cfg.bss[0] || !cfg.bss[0].ifname)
@@ -31,7 +56,7 @@ function iface_remove(cfg)
 		wdev_remove(bss.ifname);
 }
 
-function iface_gen_config(phy, config, start_disabled)
+function iface_gen_config(config, start_disabled)
 {
 	let str = `data:
 ${join("\n", config.radio.data)}
@@ -92,7 +117,7 @@ function iface_freq_info(iface, config, params)
 
 function iface_add(phy, config, phy_status)
 {
-	let config_inline = iface_gen_config(phy, config, !!phy_status);
+	let config_inline = iface_gen_config(config, !!phy_status);
 
 	let bss = config.bss[0];
 	let ret = hostapd.add_iface(`bss_config=${phy}:${config_inline}`);
@@ -123,12 +148,16 @@ function iface_config_macaddr_list(config)
 	return macaddr_list;
 }
 
-function iface_update_supplicant_macaddr(phy, config)
+function iface_update_supplicant_macaddr(phydev, config)
 {
 	let macaddr_list = [];
 	for (let i = 0; i < length(config.bss); i++)
 		push(macaddr_list, config.bss[i].bssid);
-	ubus.defer("wpa_supplicant", "phy_set_macaddr_list", { phy: phy, macaddr: macaddr_list });
+	ubus.defer("wpa_supplicant", "phy_set_macaddr_list", {
+		phy: phydev.name,
+		radio: phydev.radio ?? -1,
+		macaddr: macaddr_list
+	});
 }
 
 function __iface_pending_next(pending, state, ret, data)
@@ -143,19 +172,22 @@ function __iface_pending_next(pending, state, ret, data)
 	delete pending.defer;
 	switch (state) {
 	case "init":
-		let macaddr_list = [];
-		for (let i = 0; i < length(config.bss); i++)
-			push(macaddr_list, config.bss[i].bssid);
-		pending.call("wpa_supplicant", "phy_set_macaddr_list", { phy: phy, macaddr: macaddr_list });
+		iface_update_supplicant_macaddr(phydev, config);
 		return "create_bss";
 	case "create_bss":
-		let err = wdev_create(phy, bss.ifname, { mode: "ap" });
+		let err = phydev.wdev_add(bss.ifname, {
+			mode: "ap",
+			radio: phydev.radio,
+		});
 		if (err) {
 			hostapd.printf(`Failed to create ${bss.ifname} on phy ${phy}: ${err}`);
 			return null;
 		}
 
-		pending.call("wpa_supplicant", "phy_status", { phy: phy });
+		pending.call("wpa_supplicant", "phy_status", {
+			phy: phydev.phy,
+			radio: phydev.radio,
+		});
 		return "check_phy";
 	case "check_phy":
 		let phy_status = data;
@@ -165,12 +197,20 @@ function __iface_pending_next(pending, state, ret, data)
 
 			hostapd.printf(`Failed to bring up phy ${phy} ifname=${bss.ifname} with supplicant provided frequency`);
 		}
-		pending.call("wpa_supplicant", "phy_set_state", { phy: phy, stop: true });
+		pending.call("wpa_supplicant", "phy_set_state", {
+			phy: phydev.phy,
+			radio: phydev.radio,
+			stop: true
+		});
 		return "wpas_stopped";
 	case "wpas_stopped":
 		if (!iface_add(phy, config))
 			hostapd.printf(`hostapd.add_iface failed for phy ${phy} ifname=${bss.ifname}`);
-		pending.call("wpa_supplicant", "phy_set_state", { phy: phy, stop: false });
+		pending.call("wpa_supplicant", "phy_set_state", {
+			phy: phydev.phy,
+			radio: phydev.radio,
+			stop: false
+		});
 		return null;
 	case "done":
 	default:
@@ -185,9 +225,14 @@ function iface_pending_next(ret, data)
 	let cfg = this;
 
 	while (pending) {
-		this.next_state = __iface_pending_next(cfg, this.next_state, ret, data);
-		if (!this.next_state) {
-			__iface_pending_next(cfg, "done");
+		try {
+			this.next_state = __iface_pending_next(cfg, this.next_state, ret, data);
+			if (!this.next_state) {
+				__iface_pending_next(cfg, "done");
+				return;
+			}
+		} catch(e) {
+			hostapd.printf(`Exception: ${e}\n${e.stacktrace[0].context}`);
 			return;
 		}
 		pending = !this.defer;
@@ -229,6 +274,16 @@ function iface_pending_init(phydev, config)
 	pending.next();
 }
 
+function iface_macaddr_init(phydev, config, macaddr_list)
+{
+	let macaddr_data = {
+		num_global: config.num_global_macaddr ?? 1,
+		mbssid: config.mbssid ?? 0,
+	};
+
+	return phydev.macaddr_init(macaddr_list, macaddr_data);
+}
+
 function iface_restart(phydev, config, old_config)
 {
 	let phy = phydev.name;
@@ -246,7 +301,7 @@ function iface_restart(phydev, config, old_config)
 		return;
 	}
 
-	phydev.macaddr_init(iface_config_macaddr_list(config));
+	iface_macaddr_init(phydev, config, iface_config_macaddr_list(config));
 	for (let i = 0; i < length(config.bss); i++) {
 		let bss = config.bss[i];
 		if (bss.default_macaddr)
@@ -314,9 +369,24 @@ function bss_remove_file_fields(config)
 	return new_cfg;
 }
 
+function bss_ifindex_list(config)
+{
+	config = filter(config, (line) => !!hostapd.data.iface_fields[split(line, "=")[0]]);
+
+	return join(",", map(config, (line) => {
+		try {
+			let file = "/sys/class/net/" + split(line, "=")[1] + "/ifindex";
+			let val = trim(readfile(file));
+			return val;
+		} catch (e) {
+			return "";
+		}
+	}));
+}
+
 function bss_config_hash(config)
 {
-	return hostapd.sha1(remove_file_fields(config) + "");
+	return hostapd.sha1(remove_file_fields(config) + bss_ifindex_list(config));
 }
 
 function bss_find_existing(config, prev_config, prev_hash)
@@ -348,7 +418,7 @@ function get_config_bss(config, idx)
 	return hostapd.bss[ifname];
 }
 
-function iface_reload_config(phydev, config, old_config)
+function iface_reload_config(name, phydev, config, old_config)
 {
 	let phy = phydev.name;
 
@@ -358,13 +428,13 @@ function iface_reload_config(phydev, config, old_config)
 	if (is_equal(old_config.bss, config.bss))
 		return true;
 
-	if (hostapd.data.pending_config[phy])
+	if (hostapd.data.pending_config[name])
 		return false;
 
 	if (!old_config.bss || !old_config.bss[0])
 		return false;
 
-	let iface = hostapd.interfaces[phy];
+	let iface = hostapd.interfaces[name];
 	let iface_name = old_config.bss[0].ifname;
 	if (!iface) {
 		hostapd.printf(`Could not find previous interface ${iface_name}`);
@@ -459,7 +529,7 @@ function iface_reload_config(phydev, config, old_config)
 			return false;
 
 		let ifname = old_config.bss[i].ifname;
-		hostapd.printf(`Remove bss '${ifname}' on phy '${phy}'`);
+		hostapd.printf(`Remove bss '${ifname}' on phy '${name}'`);
 		prev_bss.delete();
 		wdev_remove(ifname);
 	}
@@ -500,11 +570,7 @@ function iface_reload_config(phydev, config, old_config)
 	}
 
 	// Step 6: assign BSSID for newly created interfaces
-	let macaddr_data = {
-		num_global: config.num_global_macaddr ?? 1,
-		mbssid: config.mbssid ?? 0,
-	};
-	macaddr_list = phydev.macaddr_init(macaddr_list, macaddr_data);
+	macaddr_list = iface_macaddr_init(phydev, config, macaddr_list);
 	for (let i = 0; i < length(config.bss); i++) {
 		if (bss_list[i])
 			continue;
@@ -528,13 +594,13 @@ function iface_reload_config(phydev, config, old_config)
 
 		let addr = phydev.macaddr_next(i);
 		if (!addr) {
-			hostapd.printf(`Failed to generate mac address for phy ${phy}`);
+			hostapd.printf(`Failed to generate mac address for phy ${name}`);
 			return false;
 		}
 		bsscfg.bssid = addr;
 	}
 
-	let config_inline = iface_gen_config(phy, config);
+	let config_inline = iface_gen_config(config);
 
 	// Step 7: fill in the gaps with new interfaces
 	for (let i = 0; i < length(config.bss); i++) {
@@ -544,17 +610,17 @@ function iface_reload_config(phydev, config, old_config)
 		if (bss)
 			continue;
 
-		hostapd.printf(`Add bss ${ifname} on phy ${phy}`);
+		hostapd.printf(`Add bss ${ifname} on phy ${name}`);
 		bss_list[i] = iface.add_bss(config_inline, i);
 		if (!bss_list[i]) {
-			hostapd.printf(`Failed to add new bss ${ifname} on phy ${phy}`);
+			hostapd.printf(`Failed to add new bss ${ifname} on phy ${name}`);
 			return false;
 		}
 	}
 
 	// Step 8: update interface bss order
 	if (!iface.set_bss_order(bss_list)) {
-		hostapd.printf(`Failed to update BSS order on phy '${phy}'`);
+		hostapd.printf(`Failed to update BSS order on phy '${name}'`);
 		return false;
 	}
 
@@ -585,7 +651,7 @@ function iface_reload_config(phydev, config, old_config)
 		if (is_equal(config.bss[i], bss_list_cfg[i]))
 			continue;
 
-		hostapd.printf(`Reload config for bss '${config.bss[0].ifname}' on phy '${phy}'`);
+		hostapd.printf(`Reload config for bss '${config.bss[0].ifname}' on phy '${name}'`);
 		if (bss.set_config(config_inline, i) < 0) {
 			hostapd.printf(`Failed to set config for bss ${ifname}`);
 			return false;
@@ -595,35 +661,36 @@ function iface_reload_config(phydev, config, old_config)
 	return true;
 }
 
-function iface_set_config(phy, config)
+function iface_set_config(name, config)
 {
-	let old_config = hostapd.data.config[phy];
+	let old_config = hostapd.data.config[name];
 
-	hostapd.data.config[phy] = config;
+	hostapd.data.config[name] = config;
 
 	if (!config) {
-		hostapd.remove_iface(phy);
+		hostapd.remove_iface(name);
 		return iface_remove(old_config);
 	}
 
-	let phydev = phy_open(phy);
+	let phy = config.phy;
+	let phydev = phy_open(phy, config.radio_idx);
 	if (!phydev) {
 		hostapd.printf(`Failed to open phy ${phy}`);
 		return false;
 	}
 
 	try {
-		let ret = iface_reload_config(phydev, config, old_config);
+		let ret = iface_reload_config(name, phydev, config, old_config);
 		if (ret) {
-			iface_update_supplicant_macaddr(phy, config);
-			hostapd.printf(`Reloaded settings for phy ${phy}`);
+			iface_update_supplicant_macaddr(phydev, config);
+			hostapd.printf(`Reloaded settings for phy ${name}`);
 			return 0;
 		}
 	} catch (e) {
-			hostapd.printf(`Error reloading config: ${e}\n${e.stacktrace[0].context}`);
+		hostapd.printf(`Error reloading config: ${e}\n${e.stacktrace[0].context}`);
 	}
 
-	hostapd.printf(`Restart interface for phy ${phy}`);
+	hostapd.printf(`Restart interface for phy ${name}`);
 	let ret = iface_restart(phydev, config, old_config);
 
 	return ret;
@@ -642,13 +709,18 @@ function config_add_bss(config, name)
 	return bss;
 }
 
-function iface_load_config(filename)
+function iface_load_config(phy, radio, filename)
 {
 	let f = open(filename, "r");
 	if (!f)
 		return null;
 
+	if (radio < 0)
+		radio = null;
+
 	let config = {
+		phy,
+		radio_idx: radio,
 		radio: {
 			data: []
 		},
@@ -675,7 +747,7 @@ function iface_load_config(filename)
 
 		if (val[0] == "#num_global_macaddr" ||
 		    val[0] == "mbssid")
-			config[val[0]] = int(val[1]);
+			config[substr(val[0], 1)] = int(val[1]);
 
 		push(config.radio.data, line);
 	}
@@ -723,16 +795,39 @@ function ex_wrap(func) {
 	};
 }
 
+function phy_name(phy, radio)
+{
+	if (!phy)
+		return null;
+
+	if (radio != null && radio >= 0)
+		phy += "." + radio;
+
+	return phy;
+}
+
+function bss_config(bss_name) {
+	for (let phy, config in hostapd.data.config) {
+		if (!config)
+			continue;
+
+		for (let bss in config.bss)
+			if (bss.ifname == bss_name)
+				return [ config, bss ];
+	}
+}
+
 let main_obj = {
 	reload: {
 		args: {
 			phy: "",
+			radio: 0,
 		},
 		call: ex_wrap(function(req) {
-			let phy_list = req.args.phy ? [ req.args.phy ] : keys(hostapd.data.config);
+			let phy_list = req.args.phy ? [ phy_name(req.args.phy, req.args.radio) ] : keys(hostapd.data.config);
 			for (let phy_name in phy_list) {
 				let phy = hostapd.data.config[phy_name];
-				let config = iface_load_config(phy.orig_file);
+				let config = iface_load_config(phy.phy, radio, phy.orig_file);
 				iface_set_config(phy_name, config);
 			}
 
@@ -742,6 +837,7 @@ let main_obj = {
 	apsta_state: {
 		args: {
 			phy: "",
+			radio: 0,
 			up: true,
 			frequency: 0,
 			sec_chan_offset: 0,
@@ -749,10 +845,10 @@ let main_obj = {
 			csa_count: 0,
 		},
 		call: ex_wrap(function(req) {
-			if (req.args.up == null || !req.args.phy)
+			let phy = phy_name(req.args.phy, req.args.radio);
+			if (req.args.up == null || !phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
-			let phy = req.args.phy;
 			let config = hostapd.data.config[phy];
 			if (!config || !config.bss || !config.bss[0] || !config.bss[0].ifname)
 				return 0;
@@ -788,10 +884,11 @@ let main_obj = {
 	},
 	config_get_macaddr_list: {
 		args: {
-			phy: ""
+			phy: "",
+			radio: 0,
 		},
 		call: ex_wrap(function(req) {
-			let phy = req.args.phy;
+			let phy = phy_name(req.args.phy, req.args.radio);
 			if (!phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
@@ -810,28 +907,34 @@ let main_obj = {
 	config_set: {
 		args: {
 			phy: "",
+			radio: 0,
 			config: "",
 			prev_config: "",
 		},
 		call: ex_wrap(function(req) {
 			let phy = req.args.phy;
+			let radio = req.args.radio;
+			let name = phy_name(phy, radio);
 			let file = req.args.config;
 			let prev_file = req.args.prev_config;
 
 			if (!phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
-			if (prev_file && !hostapd.data.config[phy]) {
-				let config = iface_load_config(prev_file);
+			if (prev_file && !hostapd.data.config[name]) {
+				let config = iface_load_config(phy, radio, prev_file);
 				if (config)
 					config.radio.data = [];
-				hostapd.data.config[phy] = config;
+				hostapd.data.config[name] = config;
 			}
 
-			let config = iface_load_config(file);
+			let config = iface_load_config(phy, radio, file);
+
+			hostapd.printf(`Set new config for phy ${name}: ${file}`);
+			iface_set_config(name, config);
 
-			hostapd.printf(`Set new config for phy ${phy}: ${file}`);
-			iface_set_config(phy, config);
+			if (hostapd.data.auth_obj)
+				hostapd.data.auth_obj.notify("reload", { phy, radio });
 
 			return {
 				pid: hostapd.getpid()
@@ -867,10 +970,40 @@ let main_obj = {
 			return 0;
 		})
 	},
+	bss_info: {
+		args: {
+			iface: ""
+		},
+		call: ex_wrap(function(req) {
+			if (!req.args.iface)
+				return libubus.STATUS_INVALID_ARGUMENT;
+
+			let config = bss_config(req.args.iface);
+			if (!config)
+				return libubus.STATUS_NOT_FOUND;
+
+			let bss = config[1];
+			config = config[0];
+			let ret = {};
+
+			for (let line in [ ...config.radio.data, ...bss.data ]) {
+				let fields = split(line, "=", 2);
+				let name = fields[0];
+				if (hostapd.data.bss_info_fields[name])
+					ret[name] = fields[1];
+			}
+
+			return ret;
+		})
+	},
 };
 
 hostapd.data.ubus = ubus;
 hostapd.data.obj = ubus.publish("hostapd", main_obj);
+
+
+let auth_obj = {};
+hostapd.data.auth_obj = ubus.publish("hostapd-auth", auth_obj);
 hostapd.udebug_set("hostapd", hostapd.data.ubus);
 
 function bss_event(type, name, data) {
@@ -885,17 +1018,45 @@ function bss_event(type, name, data) {
 return {
 	shutdown: function() {
 		for (let phy in hostapd.data.config)
-			iface_set_config(phy, null);
+			iface_set_config(phy);
 		hostapd.udebug_set(null);
 		hostapd.ubus.disconnect();
 	},
-	bss_add: function(name, obj) {
+	bss_create: function(phy, name, obj) {
+		phy = hostapd.data.config[phy];
+		if (!phy)
+			return;
+
+		if (phy.radio_idx != null && phy.radio_idx >= 0)
+			wdev_set_radio_mask(name, 1 << phy.radio_idx);
+	},
+	bss_add: function(phy, name, obj) {
 		bss_event("add", name);
 	},
-	bss_reload: function(name, obj, reconf) {
+	bss_reload: function(phy, name, obj, reconf) {
 		bss_event("reload", name, { reconf: reconf != 0 });
 	},
-	bss_remove: function(name, obj) {
+	bss_remove: function(phy, name, obj) {
 		bss_event("remove", name);
-	}
+	},
+	sta_auth: function(iface, sta) {
+		let msg = { iface, sta };
+		let ret = {};
+		let data_cb = (type, data) => {
+			ret = { ...ret, ...data };
+		};
+		if (hostapd.data.auth_obj)
+			hostapd.data.auth_obj.notify("sta_auth", msg, data_cb, null, null, 1000);
+		return ret;
+	},
+	sta_connected: function(iface, sta, data) {
+		let msg = { iface, sta, ...data };
+		let ret = {};
+		let data_cb = (type, data) => {
+			ret = { ...ret, ...data };
+		};
+		if (hostapd.data.auth_obj)
+			hostapd.data.auth_obj.notify("sta_connected", msg, data_cb, null, null, 1000);
+		return ret;
+	},
 };
diff --git a/package/network/services/hostapd/files/wpa_supplicant.uc b/package/network/services/hostapd/files/wpa_supplicant.uc
index 1709bb019e31..fbea27628e34 100644
--- a/package/network/services/hostapd/files/wpa_supplicant.uc
+++ b/package/network/services/hostapd/files/wpa_supplicant.uc
@@ -37,7 +37,7 @@ function iface_start(phydev, iface, macaddr_list)
 
 	wpas.data.iface_phy[ifname] = phy;
 	wdev_remove(ifname);
-	let ret = wdev_create(phy, ifname, wdev_config);
+	let ret = phydev.wdev_add(ifname, wdev_config);
 	if (ret)
 		wpas.printf(`Failed to create device ${ifname}: ${ret}`);
 	wdev_set_up(ifname, true);
@@ -61,22 +61,29 @@ function iface_cb(new_if, old_if)
 		iface_stop(old_if);
 }
 
-function prepare_config(config)
+function prepare_config(config, radio)
 {
 	config.config_data = readfile(config.config);
 
-	return { config: config };
+	return { config };
 }
 
-function set_config(phy_name, config_list)
+function set_config(config_name, phy_name, radio, num_global_macaddr, config_list)
 {
-	let phy = wpas.data.config[phy_name];
+	let phy = wpas.data.config[config_name];
+
+	if (radio < 0)
+		radio = null;
 
 	if (!phy) {
 		phy = vlist_new(iface_cb, false);
-		wpas.data.config[phy_name] = phy;
+		phy.name = phy_name;
+		wpas.data.config[config_name] = phy;
 	}
 
+	phy.radio = radio;
+	phy.num_global_macaddr = num_global_macaddr;
+
 	let values = [];
 	for (let config in config_list)
 		push(values, [ config.iface, prepare_config(config) ]);
@@ -92,30 +99,43 @@ function start_pending(phy_name)
 	if (!phy || !phy.data)
 		return;
 
-	let phydev = phy_open(phy_name);
+	let phydev = phy_open(phy.name, phy.radio);
 	if (!phydev) {
 		wpas.printf(`Could not open phy ${phy_name}`);
 		return;
 	}
 
 	let macaddr_list = wpas.data.macaddr_list[phy_name];
-	phydev.macaddr_init(macaddr_list);
+	phydev.macaddr_init(macaddr_list, { num_global: phy.num_global_macaddr });
 
 	for (let ifname in phy.data)
 		iface_start(phydev, phy.data[ifname]);
 }
 
+function phy_name(phy, radio)
+{
+	if (!phy)
+		return null;
+
+	if (radio != null && radio >= 0)
+		phy += "." + radio;
+
+	return phy;
+}
+
 let main_obj = {
 	phy_set_state: {
 		args: {
 			phy: "",
+			radio: 0,
 			stop: true,
 		},
 		call: function(req) {
-			if (!req.args.phy || req.args.stop == null)
+			let name = phy_name(req.args.phy, req.args.radio);
+			if (!name || req.args.stop == null)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
-			let phy = wpas.data.config[req.args.phy];
+			let phy = wpas.data.config[name];
 			if (!phy)
 				return libubus.STATUS_NOT_FOUND;
 
@@ -124,7 +144,7 @@ let main_obj = {
 					for (let ifname in phy.data)
 						iface_stop(phy.data[ifname]);
 				} else {
-					start_pending(req.args.phy);
+					start_pending(name);
 				}
 			} catch (e) {
 				wpas.printf(`Error chaging state: ${e}\n${e.stacktrace[0].context}`);
@@ -136,10 +156,11 @@ let main_obj = {
 	phy_set_macaddr_list: {
 		args: {
 			phy: "",
+			radio: 0,
 			macaddr: [],
 		},
 		call: function(req) {
-			let phy = req.args.phy;
+			let phy = phy_name(req.args.phy, req.args.radio);
 			if (!phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
@@ -149,13 +170,15 @@ let main_obj = {
 	},
 	phy_status: {
 		args: {
-			phy: ""
+			phy: "",
+			radio: 0,
 		},
 		call: function(req) {
-			if (!req.args.phy)
+			let phy = phy_name(req.args.phy, req.args.radio);
+			if (!phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
-			let phy = wpas.data.config[req.args.phy];
+			phy = wpas.data.config[phy];
 			if (!phy)
 				return libubus.STATUS_NOT_FOUND;
 
@@ -185,20 +208,23 @@ let main_obj = {
 	config_set: {
 		args: {
 			phy: "",
+			radio: 0,
+			num_global_macaddr: 0,
 			config: [],
 			defer: true,
 		},
 		call: function(req) {
-			if (!req.args.phy)
+			let phy = phy_name(req.args.phy, req.args.radio);
+			if (!phy)
 				return libubus.STATUS_INVALID_ARGUMENT;
 
-			wpas.printf(`Set new config for phy ${req.args.phy}`);
+			wpas.printf(`Set new config for phy ${phy}`);
 			try {
 				if (req.args.config)
-					set_config(req.args.phy, req.args.config);
+					set_config(phy, req.args.phy, req.args.radio, req.args.num_global_macaddr, req.args.config);
 
 				if (!req.args.defer)
-					start_pending(req.args.phy);
+					start_pending(phy);
 			} catch (e) {
 				wpas.printf(`Error loading config: ${e}\n${e.stacktrace[0].context}`);
 				return libubus.STATUS_INVALID_ARGUMENT;
@@ -242,6 +268,33 @@ let main_obj = {
 			return 0;
 		}
 	},
+	bss_info: {
+		args: {
+			iface: "",
+		},
+		call: function(req) {
+			let ifname = req.args.iface;
+			if (!ifname)
+				return libubus.STATUS_INVALID_ARGUMENT;
+
+			let iface = wpas.interfaces[ifname];
+			if (!iface)
+				return libubus.STATUS_NOT_FOUND;
+
+			let status = iface.ctrl("STATUS");
+			if (!status)
+				return libubus.STATUS_NOT_FOUND;
+
+			let ret = {};
+			status = split(status, "\n");
+			for (let line in status) {
+				line = split(line, "=", 2);
+				ret[line[0]] = line[1];
+			}
+
+			return ret;
+		}
+	},
 };
 
 wpas.data.ubus = ubus;
-- 
2.34.1

